gendlib.3
\."
\." $Id: gendlib.3,v 1.5 2003-04-07 20:05:33 achu Exp $
\." $Source: /g/g0/achu/temp/genders-cvsbackup-full/genders/gendlib.3,v $
\." 
.\" Copyright (C) 2000-2001 Regents of the University of California
.\" See the DISCLAIMER file distributed with this package.
\."
.TH GENDLIB 3 "10/10/97" "LLNL" "GENDLIB"
.SH NAME
gendlib \- perl library for querying genders file
.SH SYNOPSIS
require ``/usr/lib/genders/gendlib.pl'';
.SH DESCRIPTION
This package contains common functions for manipulating the genders file.
Previously this code existed only in the nodeattr command, but applications
that must repeatedly call nodeattr would spend a great deal of time reading
and parsing the genders file.  These functions create an in-memory
cache of genders on the first call;  subsequent calls need
only reference the in-memory copy and are thus speeded up.
.LP
.B Initialization:
.LP
Genders::init([$path_genders])
.IP
Function normally called internally for initialization.  It can be called
externally with a non-default genders file path and subsequent operations
will use this non-default data.  May be called more than once to change
genders files.  
.LP
.B Simple Queries:
.LP
Genders::hasattr($attribute, [$node])
.IP
Return 1 if node has attribute, 0 if not.  If node is not specified,
use local node.
.LP
Genders::getattrval($attribute, [$node])
.IP
Return value of attribute held by node, or empty string if no value or
node does not have attribute.  If node is not specified, use local node.
.LP
Genders::getnode($attribute)
.IP
Return a list of nodes having the specified attribute.  If a value is 
also specified ("attr=val"), only nodes with the specified attribute and
value are returned.
.LP
Genders::getattr([$node])
.IP
Return a list of attributes held by node.  If node is not specified, use
local node.
.LP
.B Complex Queries:
.LP
Genders::getallattr()
.IP
Return a list of all attributes in the genders file (one cluster only).
.LP
Genders::get_node_hash(\\%node)
.IP
Get a copy of hash of attributes to node lists for the current cluster
(a "hash of lists").  Ensure that keys exist for all possible attributes 
across clusters (though they may point to empty lists).
.LP
Genders::get_clusters()
.IP
Return a list with the local cluster name as the first and only
element.  get_clusters() returns a list instead of a scalar for
backwards compatability.  If the local cluster name is not known, null
is returned.
.LP
Genders::gendexp($exp, [$node])
.IP
Evaluate expression involving genders attributes and return the result
of the evaluation.  Any legal perl expresion using numeric constants,
genders attributes (which are converted into $variables), and the following
operators is valid:  !, ||, &&, *, +, -, /.  If $node is not specified,
assume the local host.
.LP
.B Conversion:
.LP
Genders::to_altnames()
.IP
Given a list of hostnames as they appear in the genders file, return
a list of alternate names.  On the SP, genders names are assumed to be the 
initial_hostnames, and we use the SDR to convert to reliable_hostnames.  
On other systems, we look for the value of the altname attribute in the 
genders file.
.LP
Genders::to_gendnames()
.IP
Performs the inverse of the to_altnames() function.
.SH "SEE ALSO"
nodeattr(1), libgenders(3)


NEWS
This file describes changes in recent versions of genders. It
primarily documents those changes that are of interest to users
and admins.

Changes in genders-1.4-1
===========================
 -- Added internal indexing and indexing API functions to increase
    search speed for genders.
 -- Added genders querying ability.
 -- Removed maximum attribute and value limitations.

Changes in genders-1.3-1
===========================
 -- Added fixes for the following corner cases:
    - duplicate attributes
    - whitespace between attributes and inline comments
    - trailing whitespace
    - no nodes listed
    - no attributes listed
 -- Added hostrange nodename support to specify ranges of hosts
    rather than a single node on each line of the genders file
 -- Nodes can now be specified on multiple lines
 -- Re-ordered error return code priority within many API functions.
    For example, when an array of length 0 is specified,
    GENDERS_ERR_OVERFLOW may be returned instead of GENDERS_ERR_PARAMETERS.
 
Changes in genders-1.2-1
===========================
 -- Removed implicit "all" attribute from compat/gendlib.pl.  If
    future use of the "all" attribute is required, it must be listed
    as an attribute of every node in the genders database.  See
    the genders.sample file for an example.
 -- Removed use of the /etc/clusters file from compat/gendlib.pl.  If
    future use of a cluster name is required, it must be listed as the
    value of the "cluster" attribute for each node in the genders
    database.  See the genders.sample file for an example.
 -- Removed -r and -C options from nodeattr.  If future use of the
    cluster name is required, the value of the "cluster" attribute can
    be retrieved using the -v option.

Changes in genders-1.2-pre1
===========================
 -- Hostlist code in hostlist.pl (genders-compat package only)
    can now expand ranges without a prefix (.e.g "[0-10]" and "0-10")
    as well as compress back to this same form.
 -- started NEWS file.

$Id: NEWS,v 1.16 2005-01-20 00:54:42 achu Exp $


gendlib.pl
#############################################################################
#  $Id: gendlib.pl,v 1.9 2003-04-08 15:57:59 achu Exp $
#############################################################################
#  Copyright (C) 2001-2002 The Regents of the University of California.
#  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
#  Written by Jim Garlick <garlick@llnl.gov>.
#  UCRL-CODE-2003-004.
#  
#  This file is part of Genders, a cluster configuration database and
#  rdist preprocessor.
#  For details, see <http://www.llnl.gov/linux/genders/>.
#  
#  Genders is free software; you can redistribute it and/or modify it under
#  the terms of the GNU General Public License as published by the Free
#  Software Foundation; either version 2 of the License, or (at your option)
#  any later version.
#  
#  Genders is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
#  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
#  details.
#  
#  You should have received a copy of the GNU General Public License along
#  with Genders; if not, write to the Free Software Foundation, Inc.,
#  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
#############################################################################

package Genders;

use strict;
use vars qw($included $debug $havePSSP $handle $path_genders);
use vars qw($init_called $init_hname_called $hname);
use vars qw($altAttr $clusterAttr $allAttr);
use vars qw($GENDERS_ERR_SUCCESS $GENDERS_ERR_NOTFOUND);

use lib '/usr/lib/genders/';
use Libgenders;

if (!$included) {
$included = 1;

##
## Package globals
##

$debug = 		0;

# See note in to_altnames().
$havePSSP = 		(-d "/usr/lpp/ssp");
if ($havePSSP) {
	require "/admin/lib/sdrlib.pl";
} 

$path_genders = 	Libgenders::string_constant("DEFAULT_GENDERS_FILE");
$altAttr =              Libgenders::string_constant("GENDERS_ALTNAME_ATTRIBUTE");
$clusterAttr =          Libgenders::string_constant("GENDERS_CLUSTER_ATTRIBUTE");
$allAttr =              Libgenders::string_constant("GENDERS_ALL_ATTRIBUTE");

$hname =		""; 	# short hostname

$init_called = 0;               # needed for self-initialization
$init_hname_called = 0;

$handle = 0;                    # genders handle

$GENDERS_ERR_SUCCESS =  Libgenders::constant("GENDERS_ERR_SUCCESS", 0);   
$GENDERS_ERR_NOTFOUND =  Libgenders::constant("GENDERS_ERR_NOTFOUND", 0);   

##
## Subroutines
##

# initialize package
#	$path (IN)	[optional] path to genders file
#	$rv (RETURN)	0 on failure opening genders, 1 on success
sub init
{
        my (@alist, $attr, $node, $blob, $gendfile, $ret);
        if (@_) {
	        $gendfile = $_[0];
	} else {	
	        $gendfile = ($path_genders);
	}
    
        if ($init_called) {
	        # destroy previous handle
	        Libgenders::genders_close($handle);
	        Libgenders::genders_handle_destroy($handle);
	        $handle = 0;  
	      
	        $init_called = 0;
	}
    
	$handle = Libgenders::genders_handle_create();
	if ($handle == 0) {
	        $debug && print "Error, genders_handle_create()\n";
	        return 0;
	}

	$ret = Libgenders::genders_open($handle, "");
	if ($ret == -1) {
	        $debug && print "Error, genders_open()\n";
		Libgenders::genders_handle_destroy($handle);
		$handle = 0;  
	        return 0;
	}
	
	# initialize hostname
	if (!$init_hname_called) {
	        $ret = init_hname();
	        if ($ret == 0) {
		        Libgenders::genders_close($handle);
 		        Libgenders::genders_handle_destroy($handle);
		        $handle = 0;  
		        return 0;
	      }
	}

	$init_called = 1;
	$debug && print("init called\n");
    
	return 1;
}

# cache the local hostname
sub init_hname
{
        # get 'my' hostname
        $hname = Libgenders::genders_getnodename($handle);
	if ($hname eq "") {
	        print "Error, genders_getnodename()\n";
	        return 0;
	}

	$init_hname_called = 1;
	$debug && print("init_hname called\n");

	return 1;
}


# determine if host has attribute
#	$attr (IN)	attribute
#	$node (IN)	[optional] hostname
#	$found (RETURN)	0 if not found, 1 if found
sub hasattr
{
        if (!$init_called) {
	        init();
	        if (!$init_called) {
		        return 0;
		}
	}

	my $attr = shift(@_);
	my $node = (@_) ? shift(@_) : $hname;

	my ($ret);

	if (defined($attr)) {
	        $ret = Libgenders::genders_testattr($handle, $node, $attr);
	        if ($ret == -1) {
	   	        $debug && print "Error, genders_testattr()\n";    
		        return 0;
		} 
		else {
		        return $ret;
		}
	}
	
	return 0;
}

# return value of attribute held by host
#	$attr (IN)	attribute
#	$node (IN)	[optional] hostname
#	$value (RETURN)	value or null if (no value or node does not have attr)
sub getattrval
{
        if (!$init_called) {
	        init();
		if (!$init_called) {
		        return "";
		}
	}
	
	my $attr = shift(@_);
	my $node = (@_) ? shift(@_) : $hname;

	my ($val);
	
	if (defined($attr)) {
	        $val = Libgenders::genders_getattrval($handle, $node, $attr);
		if ($val eq "" && 
		    Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
		        $debug && print "Error, genders_getattrval()\n";    
		        return "";
		} 
		else {
		        return $val;
		}
	}	    
	return "";
}

# get list of attributes held by node
#	$node (IN)	[optional] hostname
#	@attrs (RETURN)	list of attributes
sub getattr
{
        if (!$init_called) {
	        init();
		if (!$init_called) {
		        return ();
		}
	}
    
	my $node = (@_) ? shift(@_) : $hname;
    
	my ($attr);
	
	$attr = Libgenders::genders_getattr($handle, $node);
	if (@$attr == 0 &&
	    Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
	        $debug && print "Error, genders_getattr()\n";
	        return ();
	}
	else {
	        return @$attr;
	}
}

# get all attributes in genders file
#	@attrs (RETURN)	list of attributes
sub getallattr
{
        if (!$init_called) {
	        init();
		if (!$init_called) {
		        return ();
		}
	}
    
	my ($attr);
	
	$attr = Libgenders::genders_getattr_all($handle);
	if (@$attr == 0 &&
	    Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
	        $debug && print "Error, genders_getattr_all()\n";
		return ();
	}
	else {
	        return @$attr;
	}
}

# get list of nodes that have attribute
#	$attr (IN)	attribute
#	@nodes (RETURN)	list of nodes
sub getnode
{
        if (!$init_called) {
	        init();
		if (!$init_called) {
		        return ();
		}
	}

	my $attr = shift(@_);
	my (@attrtemp, $attrname, $attrval, $nodes);

	if (defined($attr)) {
	        if ($attr =~ /=/) {
		        #strip attribute name and value
		        @attrtemp = split(/=/, $attr);
			if (@attrtemp != 2) {
			        return ();
			}
			else {
			        $attrname = $attrtemp[0];
			        $attrval = $attrtemp[1];
			}
		}
		else {
		        $attrname = $attr;
		        $attrval = "";
		}
	}
	else {
	    $attrname = "";
	    $attrval = "";
	}
	
	$nodes = Libgenders::genders_getnodes($handle, $attrname, $attrval);
	if (@$nodes == 0 &&
	        Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
	        $debug && print "Error, genders_getnodes()\n";
		return ();
	}
	else {
	        return @$nodes;
	}
}

# Get a copy of hash of attributes -> node lists.  
# 	\%nodes (OUT)  	node hash
sub get_node_hash
{
        my ($nodes) = (@_);

        my (%nodes, $attrs, $attr, $attrnodes);

	if (!$init_called) {
	        init();
		if (!$init_called) {
		        %{$nodes} = ();
		}
	}

	if ($init_called) {
	    # must construct hash
	    $attrs = Libgenders::genders_getattr_all($handle);
	    if (@$attrs == 0 &&
		Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
		$debug && print "Error, genders_getattr_all()\n";
		$nodes = ();
	    }
	    elsif (@$attrs == 0) {
		%nodes = ();
	    }
	    else {
		foreach $attr (@$attrs) {
		    $attrnodes = Libgenders::genders_getnodes($handle, $attr, "");
		    if (@$attrnodes == 0 &&
			Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
			$debug && print "Error, genders_getnodes()\n";
			%nodes = ();
			last;
		    }
		    else {
			@{$nodes{$attr}} = @$attrnodes;   
		    }
		}
	    }
	}
	%{$nodes} = %nodes;
}

# initialize list of clusters
#	$path (IN)	[optional] path to clusters file
#	$rv (RETURN)	0 on failure opening attributes, 1 on success
sub init_clusters
{
        # clusters file now removed, just return 1
        return(1);
}

# get a copy of the list of clusters
#       $rv (RETURN)    "" on failure, cluster name on success
sub get_clusters
{
        if (!$init_called) {
 	        init();
		if (!$init_called) {
		        return "";
		}
	}

        my ($cluster);

	$cluster = Libgenders::genders_getattrval($handle, "", $clusterAttr);
	if ($cluster eq "" && 
	    Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
	        $debug && print "Error, genders_getattrval()\n";    
	        return "";
	} 

	return $cluster;
}


# evaluate expression involving genders attributes
# (assume any non-numeric token is a genders attribute;  turn this into
# a boolean variable with value assigned by hasattr(), then evaluate
# the expression and return the result)
# 	$exp (IN)	expression to evaluate
#	$node (IN)	[optional] hostname, use local host if unspecified
#	$rv (RETURN)	result of expression evaluation
sub gendexp
{
	if (!$init_called) {
		init();
		if (!$init_called) {
		    return 0;
		}
	}

	my $exp = shift(@_);
	my $node = (@_) ? shift(@_) : $hname; 

	my ($var, $pro, %vars, $ret);

	if (!$exp) {
		return $exp;
	}

	$pro = "";
	foreach $var (split(/[\!\+\-\*\/(\s\(\&\|)]+/, $exp)) {
		$var =~ s/\s+//g;
		next if (!$var || $var =~ /^[0-9]+$/);
		$ret = Libgenders::genders_testattr($handle, $node, $var);
		if ($ret == -1) {
	   	        $debug && print "Error, genders_testattr()\n";    
		        return 0;
		} 
		elsif ($ret) {
			$pro .= "my \$$var = 1; ";
		} else {
			$pro .= "my \$$var = 0; ";
		}
		$vars{$var}++;
	}
	foreach $var (keys %vars) {
		$exp =~ s/$var/\$$var/g;
	}
	if ($debug) {
		printf("evaluating { %s } for host %s\n", $pro . $exp, $node);
	}
	no strict;
	return(eval($pro . $exp));
	use strict;
}

# Convert "genders names" to an alternate names.  On an SP, the genders
# name is assumed to be the SDR initial_hostname, and the alternate name
# is the reliable_hostname.  On other systems, the alternate hostname is
# stored in the genders for each node as the value of the "altname" attribute.
#	@inList (IN)	list of genders names
#	RETURN		list of alternate names
# NOTE: names in the input that cannot be converted are preserved in the output
sub to_altnames
{
	if (!$init_called) {
		init();
		if (!$init_called) {
		    return ();
		}
	}

	my (@inList) = @_;
	my (@outList, $altName, $name);

	foreach $name (@inList) {
		($name) = split(/\./, $name);	# shorten name
		if ($havePSSP) {
			$altName = Sdr::nn2ename(Sdr::sname2nn($name));
		} else {
		    $altName = Libgenders::genders_getattrval($handle, $name, $altAttr);
		    if ($altName eq "" && 
			Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS &&
			Libgenders::genders_errnum($handle) != $GENDERS_ERR_NOTFOUND) {
			$debug && print "Error, genders_getattrval()\n";    
			return ();
		    } 

		}
		push(@outList, $altName ? $altName : $name);
	}	
	return(@outList);
}

# perform the inverse of to_altnames()
#	@inList (IN)	list of altnernate names
#	RETURN		list of genders names
# NOTE: names in the input that cannot be converted are preserved in the output
sub to_gendnames
{
	if (!$init_called) {
		init();
		if (!$init_called) {
		    return ();
		}
	}

	my (@inList) = @_;
	my (@outList, $altName, $name, $tmp, $nodes, $val);

	foreach $altName (@inList) {
	        $name = "";
		($altName) = split(/\./, $altName);	# shorten name
		if ($havePSSP) {
			$name = Sdr::nn2sname(Sdr::ename2nn($altName));
		} else {
		    $nodes = Libgenders::genders_getnodes($handle, $altAttr, "");
		    if (@$nodes == 0 &&
			Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
			$debug && print "Error, genders_getnodes()\n";    
			return ();
		    }

		    foreach $tmp (@$nodes) {
			$val = Libgenders::genders_getattrval($handle, $tmp, $altAttr);
			if ($val eq "" && 
			    Libgenders::genders_errnum($handle) != $GENDERS_ERR_SUCCESS) {
			    $debug && print "Error, genders_getattrval()\n";    
			    return ();
			} 
			
			if ($val eq $altName) {
			    $name = $tmp;
			    last;
			}
		    }
		}
		push(@outList, $name ? $name : $altName);
	}	
	return(@outList);
}

}	# $included
1;	# return a true value...
